programs/install/Install.cpp

```diff
@@ -143,57 +154,89 @@ int mainEntryClickHouseInstall(int argc, char ** argv)
+                main_bin_path.string());
+
+            if (filesEqual(binary_self_path.string(), main_bin_path.string()))
+            {
+                already_installed = true;
+                fmt::print("ClickHouse binary is already located at {} and it has the same content as {}\n",
+                    main_bin_path.string(), binary_self_canonical_path.string());
+            }
         }
 
-        if (fs::exists(main_bin_path))
+        if (already_installed)
         {
-            fmt::print("{} already exists, will rename existing binary to {} and put the new binary in place\n",
-                       main_bin_path.string(), main_bin_old_path.string());
-
-            /// There is file exchange operation in Linux but it's not portable.
-            fs::rename(main_bin_path, main_bin_old_path);
+            if (0 != chmod(main_bin_path.string().c_str(), S_IRUSR | S_IRGRP | S_IROTH | S_IXUSR | S_IXGRP | S_IXOTH))
+                throwFromErrno(fmt::format("Cannot chmod {}", main_bin_path.string()), ErrorCodes::SYSTEM_ERROR);
         }
+        else
+        {
+            size_t available_space = fs::space(bin_dir).available;
+            if (available_space < binary_size)
+                throw Exception(ErrorCodes::NOT_ENOUGH_SPACE, "Not enough space for clickhouse binary in {}, required {}, available {}.",
+                    bin_dir.string(), ReadableSize(binary_size), ReadableSize(available_space));
+
+            fmt::print("Copying ClickHouse binary to {}\n", main_bin_tmp_path.string());
 
-        fmt::print("Renaming {} to {}.\n", main_bin_tmp_path.string(), main_bin_path.string());
-        fs::rename(main_bin_tmp_path, main_bin_path);
+            try
+            {
+                ReadBufferFromFile in(binary_self_path.string());
+                WriteBufferFromFile out(main_bin_tmp_path.string());
+                copyData(in, out);
+                out.sync();
+
+                if (0 != fchmod(out.getFD(), S_IRUSR | S_IRGRP | S_IROTH | S_IXUSR | S_IXGRP | S_IXOTH))
+                    throwFromErrno(fmt::format("Cannot chmod {}", main_bin_tmp_path.string()), ErrorCodes::SYSTEM_ERROR);
+
+                out.finalize();
+            }
+            catch (const Exception & e)
+            {
+                if (e.code() == ErrorCodes::CANNOT_OPEN_FILE && geteuid() != 0)
+                    std::cerr << "Install must be run as root: sudo ./clickhouse install\n";
+                throw;
+            }
+
+            if (old_binary_exists)
+            {
+                fmt::print("{} already exists, will rename existing binary to {} and put the new binary in place\n",
+                        main_bin_path.string(), main_bin_old_path.string());
+
+                /// There is file exchange operation in Linux but it's not portable.
+                fs::rename(main_bin_path, main_bin_old_path);
+            }
+
```

Comment:
And for how long / why 'normal' users would store .old file?..

I guess they will not be even aware that it exists / can be used somehow...

At the same time admins prefers to keep /usr/bin free from backups / unused stuff.
Also old version can be potentially unsafe (have some dangerous bug).

I would prefer to disable that behavior with some flag when clickhouse is installed via deb / rpm. (there is a standard way to rollback in those cases).

Reply:
I also have some doubts about extra disk usage (sometimes the root volume is very small).

The .old is kept just in case if quick rollback is required or if the new file is damaged after writing to disk...